diff -ur grep-2.5.3/src/dfa.c grep-2.5.3-johan/src/dfa.c
--- grep-2.5.3/src/dfa.c	2007-06-28 20:57:19.000000000 +0200
+++ grep-2.5.3-johan/src/dfa.c	2008-01-05 09:46:54.000000000 +0100
@@ -1605,9 +1605,7 @@
   int *visited;
   position p, old;
 
-  MALLOC(visited, int, d->tindex);
-  for (i = 0; i < d->tindex; ++i)
-    visited[i] = 0;
+  CALLOC(visited, int, d->tindex);
 
   for (i = 0; i < s->nelem; ++i)
     if (d->tokens[s->elems[i].index] >= NOTCHAR
